samza-core/src/main/java/org/apache/samza/operators/impl/PartialJoinOperatorImpl.java

```diff
@@ -47,35 +50,41 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(PartialJoinOperatorImpl.class);
 
+  private final PartialJoinOperatorSpec<K, M, JM, RM> partialJoinOpSpec;
   private final PartialJoinFunction<K, M, JM, RM> thisPartialJoinFn;
   private final PartialJoinFunction<K, JM, M, RM> otherPartialJoinFn;
   private final long ttlMs;
-  private final int opId;
   private final Clock clock;
 
-  PartialJoinOperatorImpl(PartialJoinOperatorSpec<K, M, JM, RM> partialJoinOperatorSpec, MessageStreamImpl<M> source,
+  PartialJoinOperatorImpl(PartialJoinOperatorSpec<K, M, JM, RM> partialJoinOpSpec,
       Config config, TaskContext context, Clock clock) {
-    this.thisPartialJoinFn = partialJoinOperatorSpec.getThisPartialJoinFn();
-    this.otherPartialJoinFn = partialJoinOperatorSpec.getOtherPartialJoinFn();
-    this.ttlMs = partialJoinOperatorSpec.getTtlMs();
-    this.opId = partialJoinOperatorSpec.getOpId();
+    this.partialJoinOpSpec = partialJoinOpSpec;
+    this.thisPartialJoinFn = partialJoinOpSpec.getThisPartialJoinFn();
+    this.otherPartialJoinFn = partialJoinOpSpec.getOtherPartialJoinFn();
+    this.ttlMs = partialJoinOpSpec.getTtlMs();
     this.clock = clock;
   }
 
   @Override
-  public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
+  protected void doInit(Config config, TaskContext context) {
+    this.thisPartialJoinFn.init(config, context);
+  }
+
+  @Override
+  public Collection<RM> handleMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
     K key = thisPartialJoinFn.getKey(message);
     thisPartialJoinFn.getState().put(key, new PartialJoinMessage<>(message, clock.currentTimeMillis()));
     PartialJoinMessage<JM> otherMessage = otherPartialJoinFn.getState().get(key);
     long now = clock.currentTimeMillis();
     if (otherMessage != null && otherMessage.getReceivedTimeMs() > now - ttlMs) {
```

Comment:
Is it useful to add a `matchedKeys` metric for the join operator? Seems like a useful one to track when doing joins to get some visibility into join correctness.